TRUNK-5724: Refactor getSimilarPeople to use Lucene

fruether opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'TRUNK-5111 Replace use of deprecated isVoided' -->
<!--- 'TRUNK-JiraIssueNumber JiraIssueTitle' -->
## Description of what I changed
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->


## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->
see https://issues.openmrs.org/browse/TRUNK-5724

which is a sub-task of https://issues.openmrs.org/browse/TRUNK-5680 to make everything deployable easier and introduce smaller changes

The talk thread discussing approach logic and so on can be found here: https://talk.openmrs.org/t/soundex-search-in-lucenequeries/26068
Please have a look there before asking for my changes!


## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [X] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [X ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [X ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

ibacher commented:
@fruether Please see our [Pull Request Tips](https://wiki.openmrs.org/display/docs/Pull+Request+Tips). In particular:

1. Change the commit message to "TRUNK-5724: Refactor getSimilarPeople to use Lucene"

2. Add a link in the Jira ticket to point to this PR.

3. In Jira, click the Request Code Review button.

fruether commented:
@ibacher any quick and fast way to rename the commit?
I assume there will be further comments. Can I change the name once I have more to fix? I am sure that will happen once the review starts.

ibacher commented:
`git commit --amend -m "TRUNK-5274: Refactor getSimilarPeople to use Lucene"`. And it's not the most important thing to take care of straight-away.

fruether commented:
@ibacher I quickly fixed it. Thank you! Looking forward to more feedback or questions!

ibacher commented on api/pom.xml:
```diff
@@ -12,7 +12,7 @@
    <parent>
       <groupId>org.openmrs</groupId>
-      <artifactId>openmrs</artifactId>
+      <artifactId>openmrs</artifactId> 
```
Sorry to be really picky, but don't add a new space here

ibacher commented on api/src/main/java/org/openmrs/ConceptName.java:
```diff
@@ -13,7 +13,6 @@
 import java.util.Date;
 import java.util.HashSet;
 import java.util.Locale;
-
```
I would remove this from your commit altogether

ibacher commented on api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java:
```diff
@@ -72,6 +73,121 @@ public void setSessionFactory(SessionFactory sessionFactory) {
+		
+		LuceneQuery<PersonName> luceneQuery = personLuceneQuery.getSoundexPersonNameQuery(query, birthyear, false, gender);;
+		ListPart<Object[]> names = luceneQuery.listPartProjection(0, maxResults, "person.personId");
+		names.getList().forEach(x-> people.add(getPerson((Integer) x[0])));
```
`x ->`

ibacher commented on api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java:
```diff
@@ -72,6 +73,121 @@ public void setSessionFactory(SessionFactory sessionFactory) {
+		Set<Person> results = new HashSet<>();
+		LuceneQuery<PersonName> luceneQuery;
+		
+		if(firstName) {
```
`if (firstName) {`

ibacher commented on api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java:
```diff
@@ -84,35 +200,19 @@ public void setSessionFactory(SessionFactory sessionFactory) {
-			            "  when soundex(pname.familyName2) = soundex(:n2) then 4").append("  else 0 ").append(" end")
-			        .append(") > 6");
+			return executeSoundexTwoPersonNameQuery(names[0], names[1], birthyear, false, gender);
 		} else if (names.length == 3) {
```
Is there a reason this case isn't covered?

fruether replied:
I want to split the issue in several parts to make discussion more structured and not make a to large PR. That is why I created a accoding subtask to the main issue.

For now I focused on the case of a names.length of 1 or 2. So we can first agree that this part works. And once this PR is done I will focus on coding this case.

ibacher replied:
That's a very reasonable route to take. Just wanted to understand what is going on here.

fruether replied:
👍 
There seems like a lot is going on while re-factoring since Lucene core functionality is touched (LuceneQuery.java) which may affects other PersonName searches although the test cases do not indicate it.

And the reverse engineering of the business logic of the sql code is not too trivial: https://talk.openmrs.org/t/soundex-search-in-lucenequeries/26068/22?u=fruether

The other comments should be fixed for now

ibacher commented on api/src/test/resources/org/openmrs/api/include/PersonServiceTest-extranames.xml:
```diff
@@ -11,26 +11,26 @@
-  <patient_identifier patient_identifier_id="31" patient_id="3" identifier="4567" identifier_type="1" preferred="1" location_id="1" creator="1" date_created="2005-01-01 00:00:00.0" voided="false" uuid="2d7c3067-346a-4cc8-9f67-66bb5b149153"/>
-  <patient_identifier patient_identifier_id="21" patient_id="4" identifier="5634" identifier_type="1" preferred="1" location_id="1" creator="1" date_created="2005-01-01 00:00:00.0" voided="true" uuid="df6633a5-3896-498f-97bd-3160ffa05d0e"/>
-  <patient_identifier patient_identifier_id="11" patient_id="5" identifier="563422" identifier_type="1" preferred="1" location_id="1" creator="1" date_created="2005-01-01 00:00:00.0" voided="false" uuid="f6e43b2a-e331-496b-87f7-27b64d84a015"/>
+	<person person_id="2" gender="M" dead="false" creator="1" date_created="2005-01-01 00:00:00.0" voided="false" uuid="1716432a-4bf9-436c-baad-f03236e960e1"/>
```
I don't particularly love converting the leading white space to tabs here. Is it possible to undo it?

ibacher commented on api/src/test/java/org/openmrs/api/PersonServiceTest.java:
```diff
@@ -478,27 +485,92 @@ public void getSimilarPeople_shouldMatchSingleSearchToAnyNamePart() throws Excep
+		assertTrue(containsId(matches, 1012));
+		assertTrue(containsId(matches, 1013));
+		
+		
```
I'm not particularly a fan of having a blank line after the last executable statement in a function, but we definitely shouldn't have two.

fruether commented on api/src/main/java/org/openmrs/api/db/hibernate/PersonLuceneQuery.java:
```diff
@@ -49,7 +49,7 @@ public PersonLuceneQuery(SessionFactory sessionFactory) {
 
 	public LuceneQuery<PersonName> getPatientNameQuery(String query, boolean includeVoided, LuceneQuery<?> skipSame) {
-		return getPersonNameQuery(query, false, includeVoided, true, skipSame);
+		return   getPersonNameQuery(query, false, includeVoided, true, skipSame);
```
I will fix this space in next commit

achilep commented:
make sure that your branch is up to date before you  next commit.

fruether commented:
@achilep 
I have a small doubt since I am not so secure with GitHub.

```
Frederiks-MBP-2:openmrs-core freddy$ git pull --rebase upstream master
remote: Enumerating objects: 245, done.
remote: Counting objects: 100% (245/245), done.
remote: Total 318 (delta 244), reused 244 (delta 244), pack-reused 73
Receiving objects: 100% (318/318), 117.29 KiB | 622.00 KiB/s, done.
Resolving deltas: 100% (260/260), completed with 209 local objects.
From https://github.com/openmrs/openmrs-core
 * branch                master     -> FETCH_HEAD
   373c34bd2..b571cb5ae  master     -> upstream/master
First, rewinding head to replay your work on top of it...
Applying: TRUNK-5274: Refactor getSimilarPeople to use Lucene
```


What can I do to execute the rebase?

fruether commented:
Okay I think the commits should now be included. However, my commits are somehow now scattered before and after the merges. Any tipps how to fix this?

```
Frederiks-MBP-2:openmrs-core freddy$ git pull --rebase upstream master
From https://github.com/openmrs/openmrs-core
 * branch                master     -> FETCH_HEAD
First, rewinding head to replay your work on top of it...
Applying: TRUNK-5274: Refactor getSimilarPeople to use Lucene
Applying: TRUNK-5274: Refactor getSimilarPeople to use Lucene
Using index info to reconstruct a base tree...
M	api/pom.xml
M	api/src/main/java/org/openmrs/Person.java
M	api/src/main/java/org/openmrs/PersonName.java
M	api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java
M	api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java
M	api/src/main/java/org/openmrs/api/db/hibernate/PersonLuceneQuery.java
M	api/src/main/java/org/openmrs/api/db/hibernate/search/LuceneAnalyzerFactory.java
M	api/src/main/java/org/openmrs/api/db/hibernate/search/LuceneAnalyzers.java
M	api/src/main/java/org/openmrs/api/db/hibernate/search/LuceneQuery.java
M	api/src/main/java/org/openmrs/util/OpenmrsConstants.java
M	api/src/test/java/org/openmrs/api/PatientServiceTest.java
M	api/src/test/java/org/openmrs/api/PersonServiceTest.java
M	api/src/test/resources/org/openmrs/api/include/PersonServiceTest-names.xml
.git/rebase-apply/patch:94: trailing whitespace.
		
.git/rebase-apply/patch:116: trailing whitespace.
	 * 
.git/rebase-apply/patch:118: trailing whitespace.
	 * @param birthyear the birthyear the searched person should have 
.git/rebase-apply/patch:119: trailing whitespace.
	 * @param includeVoided true if voided person should be included 
.git/rebase-apply/patch:120: trailing whitespace.
	 * @param gender of the person to search for 
warning: squelched 91 whitespace errors
warning: 96 lines add whitespace errors.
Falling back to patching base and 3-way merge...
No changes -- Patch already applied.
Applying: TRUNK-5274: Refactor getSimilarPeople to use Lucene
Using index info to reconstruct a base tree...
M	api/pom.xml
M	api/src/main/java/org/openmrs/Person.java
M	api/src/main/java/org/openmrs/PersonName.java
M	api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java
M	api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java
M	api/src/main/java/org/openmrs/api/db/hibernate/PersonLuceneQuery.java
M	api/src/main/java/org/openmrs/api/db/hibernate/search/LuceneAnalyzerFactory.java
M	api/src/main/java/org/openmrs/api/db/hibernate/search/LuceneAnalyzers.java
M	api/src/main/java/org/openmrs/api/db/hibernate/search/LuceneQuery.java
M	api/src/main/java/org/openmrs/util/OpenmrsConstants.java
M	api/src/test/java/org/openmrs/api/PatientServiceTest.java
M	api/src/test/java/org/openmrs/api/PersonServiceTest.java
M	api/src/test/resources/org/openmrs/api/include/PersonServiceTest-names.xml
.git/rebase-apply/patch:94: trailing whitespace.
		
.git/rebase-apply/patch:116: trailing whitespace.
	 * 
.git/rebase-apply/patch:118: trailing whitespace.
	 * @param birthyear the birthyear the searched person should have 
.git/rebase-apply/patch:119: trailing whitespace.
	 * @param includeVoided true if voided person should be included 
.git/rebase-apply/patch:120: trailing whitespace.
	 * @param gender of the person to search for 
warning: squelched 91 whitespace errors
warning: 96 lines add whitespace errors.
Falling back to patching base and 3-way merge...
Auto-merging api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java
Auto-merging api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java
Auto-merging api/src/main/java/org/openmrs/PersonName.java
Auto-merging api/src/main/java/org/openmrs/Person.java
No changes -- Patch already applied.
Frederiks-MBP-2:openmrs-core freddy$ 

```

fruether commented:
> make sure that your branch is up to date before you next commit.

@achilep DONE. It is up-to date now

achilep commented:
Sorry for late reply @fruether

achilep commented:
Great job

ibacher commented on api/src/main/java/org/openmrs/api/db/hibernate/search/LuceneAnalyzerFactory.java:
```diff
@@ -58,6 +60,12 @@ public SearchMapping getSearchMapping() {
 			.param("minGramSize", "2")
 			.param("maxGramSize", "20");
+		mapping.analyzerDef(LuceneAnalyzers.SOUNDEX_ANALYZER, StandardTokenizerFactory.class)
+			.filter(ClassicFilterFactory.class) //StandardTokenizerFactory https://www.programcreek.com/java-api-examples/index.php?api=org.apache.lucene.analysis.phonetic.PhoneticFilterFactory
```
Is there any value to this comment here?

fruether replied:
The URL shows the example I used for the setup. I will remove it in the next commit

ibacher commented on api/src/main/java/org/openmrs/api/db/hibernate/search/LuceneAnalyzers.java:
```diff
@@ -26,4 +26,6 @@ private LuceneAnalyzers() {
 	
 	public static final String PHRASE_ANALYZER = "phraseAnalyzer";
+	
+	public static final String SOUNDEX_ANALYZER = "soundexExactAnalyzer";
```
"soundexExactAnalyzer" sounds like a weird descriptor to me... SOUNDEX is, by its very nature, not an exact match, unless I'm missing something?

fruether replied:
Yes makes sense.  I will change it accordingly

ibacher commented on api/src/main/java/org/openmrs/util/OpenmrsConstants.java:
```diff
@@ -339,6 +339,8 @@ public static String getOpenmrsProperty(String property) {
 	
 	public static final String GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_START = "START";
 	
+	public static final String GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_SOUNDEX = "Soundex";
```
Shouldn't this be "SOUNDEX" to match the other types here?

fruether replied:
Done

ibacher commented on api/src/main/java/org/openmrs/PersonName.java:
```diff
@@ -58,15 +58,17 @@
 			@Field(name = "givenNameStart", analyzer = @Analyzer(definition = LuceneAnalyzers.START_ANALYZER), boost = @Boost(4f)),
-			@Field(name = "givenNameAnywhere", analyzer = @Analyzer(definition = LuceneAnalyzers.ANYWHERE_ANALYZER), boost = @Boost(2f))
+			@Field(name = "givenNameAnywhere", analyzer = @Analyzer(definition = LuceneAnalyzers.ANYWHERE_ANALYZER), boost = @Boost(2f)),
+			@Field(name = "givenNameSoundex", analyzer =  @Analyzer(definition = LuceneAnalyzers.SOUNDEX_ANALYZER), boost = @Boost(16f))
```
Do we really want such a high boost for a SOUNDEX match? I would suggest the boost should be left at 1.0.

ibacher commented on api/src/main/java/org/openmrs/PersonName.java:
```diff
@@ -58,15 +58,17 @@
 			@Field(name = "middleNameStart", analyzer = @Analyzer(definition = LuceneAnalyzers.START_ANALYZER), boost = @Boost(2f)),
-			@Field(name = "middleNameAnywhere", analyzer = @Analyzer(definition = LuceneAnalyzers.ANYWHERE_ANALYZER))
+			@Field(name = "middleNameAnywhere", analyzer = @Analyzer(definition = LuceneAnalyzers.ANYWHERE_ANALYZER)),
+			@Field(name = "middleNameSoundex", analyzer =  @Analyzer(definition = LuceneAnalyzers.SOUNDEX_ANALYZER), boost = @Boost(8f))
```
Do we really want such a high boost for a SOUNDEX match? I would suggest the boost should be 0.5.

ibacher commented on api/src/main/java/org/openmrs/PersonName.java:
```diff
@@ -76,13 +78,15 @@
 			@Field(name = "familyNameExact", analyzer = @Analyzer(definition = LuceneAnalyzers.EXACT_ANALYZER), boost = @Boost(8f)),
 			@Field(name = "familyNameStart", analyzer = @Analyzer(definition = LuceneAnalyzers.START_ANALYZER), boost = @Boost(4f)),
 			@Field(name = "familyNameAnywhere", analyzer = @Analyzer(definition = LuceneAnalyzers.ANYWHERE_ANALYZER), boost = @Boost(2f)),
+			@Field(name = "familyNameSoundex", analyzer =  @Analyzer(definition = LuceneAnalyzers.SOUNDEX_ANALYZER), boost = @Boost(16f))
```
Do we really want such a high boost for a SOUNDEX match? I would suggest the boost should be left at 1.0.

fruether replied:
I can set all to 1.0. It should have no effect I think 👍

ibacher commented on api/src/main/java/org/openmrs/PersonName.java:
```diff
@@ -76,13 +78,15 @@
 			@Field(name = "familyName2Exact", analyzer = @Analyzer(definition = LuceneAnalyzers.EXACT_ANALYZER), boost = @Boost(4f)),
 			@Field(name = "familyName2Start", analyzer = @Analyzer(definition = LuceneAnalyzers.START_ANALYZER), boost = @Boost(2f)),
 			@Field(name = "familyName2Anywhere", analyzer = @Analyzer(definition = LuceneAnalyzers.ANYWHERE_ANALYZER)),
+			@Field(name = "familyName2Soundex", analyzer =  @Analyzer(definition = LuceneAnalyzers.SOUNDEX_ANALYZER), boost = @Boost(8f))
```
Do we really want such a high boost for a SOUNDEX match? I would suggest the boost should be 0.5.

ibacher commented on api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java:
```diff
@@ -787,8 +787,8 @@ public Long getCountOfPatients(String query, boolean includeVoided) {
 		PersonLuceneQuery personLuceneQuery = new PersonLuceneQuery(sessionFactory);
-
 		LuceneQuery<PersonName> nameQuery = personLuceneQuery.getPatientNameQuery(query, includeVoided, identifierQuery);
+		
```
Does this change need to be in this PR?

fruether replied:
Fixed

ibacher commented on api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java:
```diff
@@ -72,6 +73,121 @@ public void setSessionFactory(SessionFactory sessionFactory) {
+	private Set<Person> executeSoundexOnePersonNameQuery(String name, Integer birthyear, boolean includeVoided , String gender) {
+		// Use the Soundex Search Mode for each query
+		String matchMode = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_MODE);
+		Context.getAdministrationService().setGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_MODE, OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_SOUNDEX);
```
What happens when two people search for patients at the same time in different browsers?

ibacher commented on api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java:
```diff
@@ -72,6 +73,121 @@ public void setSessionFactory(SessionFactory sessionFactory) {
+	private Set<Person> executeSoundexTwoPersonNameQuery(String name1, String name2,  Integer birthyear, boolean includeVoided , String gender) {
+		// Use the Soundex Search Mode for each query
+		String matchMode = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_MODE);
+		Context.getAdministrationService().setGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_MODE, OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_SOUNDEX);
```
What happens when two people search for patients at the same time in different browsers?

fruether replied:
Honestly, I don't know. Of all the options it seems like this was the smoothest way. 
Perhaps there is a constant that is bound to a session? 
Or there is a near way to set the choosen Lucene Analyzer that I was not aware of

fruether replied:
I asked in talk if we can get some support deciding what the approach would be. Hope you do not mind!

ibacher replied:
Not at all!

ibacher commented on api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java:
```diff
@@ -72,6 +73,121 @@ public void setSessionFactory(SessionFactory sessionFactory) {
+		ListPart<Object[]> names = luceneQuery.listPartProjection(0, maxResults, "person.personId");
+		names.getList().forEach(x -> people.add(getPerson((Integer) x[0])));
+		
+		Context.getAdministrationService().setGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_MODE, matchMode);
```
What happens when two people search for patients at the same time in different browsers?

ibacher commented on api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java:
```diff
@@ -72,6 +73,121 @@ public void setSessionFactory(SessionFactory sessionFactory) {
+		firstNamesMatchSecondName.retainAll(secondNamesMatch);
+		people.addAll(firstNamesMatchSecondName);
+		
+		Context.getAdministrationService().setGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_PATIENT_SEARCH_MATCH_MODE, matchMode);
```
What happens when two people search for patients at the same time in different browsers?

fruether commented:
@ibacher Thanks for the review. I can understand your feeling regarding the Lucene componentes. The issue is basically described in this talk topic: https://talk.openmrs.org/t/soundex-search-in-lucenequeries/26068/27?u=fruether

In order to get LuceneAnalyzer to work I have to avoid this line from being executed:  	
`analyzer = getFullTextSession().getSearchFactory().getAnalyzer(LuceneAnalyzers.EXACT_ANALYZER);
`
Basically there have been the following options I am aware of:
(1) Use an existing global contant to make a switch
(2) Drop the if condition. Dropping the if conditions leads to a lot of test cases to go wrong because somehow it does effect the other searches.... 

@rkorytkowski would perhaps a person suited to answer our Lucene questions?

ibacher commented:
@fruether Thanks for the explanation. I see the difficulty now. What about the idea of passing an enum for the match type. Something like:

```java
Enum MatchType {
	EXACT,
	INEXACT
}
```

And then we store the `MatchType` as a new required field in the `LuceneQuery` class. That way, we have the type of match inside the query object and we can change that if statement to something like:

```java
if (matchType.equals(MatchType.EXACT) {
    analyzer = getFullTextSession().getSearchFactory().getAnalyzer(LuceneAnalyzers.EXACT_ANALYZER);
} else {
    analyzer = getFullTextSession().getSearchFactory().getAnalyzer(getType());
}
```

fruether commented:
@ibacher Awesome! I really like the idea. However, I would adapt it slightly. I would define an enum as following with NORMAL (may we find a better name) as default value:
```
Enum MatchType {
        NORMAL,
	SOUNDEX
}
```
No we could create a logic like:

```
if (matchType.equals(MatchType.SOUNDEX) {
}
Else 
...... as it is right now
```

The advantage of this approach would be that as less of the previous existing code is changed as possible. That should give stability. Once this passes I would like to create a talk thread and issue discussing the behavior I found regarding the failing test cases i investiged.
Then this function overall can may be made more straight forward.

What do you think about that?
Thanks for the support!

ibacher commented:
@fruether That's a fair point!

ibacher commented:
@fruether Thanks! I think that addresses all of my concerns, although I'd still need to find some time to play around with this in an actual system before I'd be comfortable committing it.

fruether commented:
@ibacher Thank you! I appreciate your time and feedback. 
I can understand that you want a more in depth test. I think it is a good approach to first test it more in depth. 
In case your are first busy with more priority tasks just let me know from time to time if you are still on it.

If that works the three paramter search can be tested less frequently

fruether commented:
The new updates to the branch were just related to drop added spaces and lines. No logic change

fruether commented:
FYI: I applied TRUNK-5492 to this node to make it mergeable

fruether commented:
@ibacher Just wondering are there any new insights regarding this PR?
I will soon push an update to the branch to resolve the merge conflict.

ibacher commented:
@fruether I'm sorry I haven't updated this recently. I've been working on getting an environment that works with the latest core changes, this PR and the RefApp, which is a bit more of a lift than it usually is. I just wanted to let you know this hasn't fallen off my radar.

fruether commented:
Alright. Great to hear that you are still on it. Hope you are doing well!

Ian <notifications@github.com> schrieb am Mo., 20. Apr. 2020, 16:07:

> @fruether <https://github.com/fruether> I'm sorry I haven't updated this
> recently. I've been working on getting an environment that works with the
> latest core changes, this PR and the RefApp, which is a bit more of a lift
> than it usually is. I just wanted to let you know this hasn't fallen off my
> radar.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-core/pull/3152#issuecomment-616576971>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMEOAH2X4LBZ4STQDI7NSDRNRJIHANCNFSM4LRMZI3A>
> .
>

fruether commented:
@ibacher hope you are doing fine! I will probably have the time during the next 14 days to work on the function for getSimilar() block with three arguments (n1,n2,n3). 

I just wanted to check if this PR is still on your radar since it may has an effect to the discussed implementation for three arguments.

ibacher commented:
@fruether Thanks for pinging me. Yes, this is still on my radar and I'm hopefully going to have a chance to run the tests I wanted to tomorrow. Sorry for the long process here. I've been fighting with getting the RefApp running on the latest 2.4. I would suggest going ahead with the `getSimilar()` for three arguments anyways as this looks like a good approach... We just might need to tweak some details here or there.

dkayiwa commented:
@ibacher as for getting the refapp to run on platform 2.4, did you try these addons? https://ci.openmrs.org/browse/REFAPP-OMODDISTRO-9879/artifact

ibacher commented:
@dkayiwa That probably would've made things easier, but I had started before that was released and spent some time fighting with my forked versions of things.

fruether commented on api/src/main/java/org/openmrs/api/db/hibernate/HibernatePersonDAO.java:
```diff
@@ -72,6 +73,109 @@ public void setSessionFactory(SessionFactory sessionFactory) {
+		int maxResults = HibernatePersonDAO.getMaximumSearchResults();
+		
+		// Execute a search on all names of a person with the given criteria
+		LuceneQuery<PersonName> luceneQuery = personLuceneQuery.getSoundexPersonNameSearchOnAllNames(name1, name2, 0, false, "M");
```
Remove this

fruether commented:
@ibacher I fixed a bug in the code. At one function call the gender was hard coded. I fixed it right now. Make sure to pull before testing on your side .

fruether commented:
@ibacher I addes the 3 name changes to this PR: https://github.com/openmrs/openmrs-core/pull/3241 in case you want to test it may be an approach that we switch the PR and I close this one? What do you think

fruether commented:
@ibacher any thing I can do to help you getting this PR or even better the successive PR through the door? 

Any timeframe? Because there is one last else branch missing that still has to be refactored. Depending on your time I may could start the last block during the next weeks and then we make one finaly/large PR?

ibacher commented:
@fruether Does the other PR include these changes? Because, if so, let's close this one and move onto that one. I'm actually quite happy with these changes.

fruether commented:
Yes. It contains the same changes. So we could close this one and focus on
the other one.
It depends if we would prefer one large commit or multiple smaller once

Ian <notifications@github.com> schrieb am Mo., 22. Juni 2020, 14:54:

> @fruether <https://github.com/fruether> Does the other PR include these
> changes? Because, if so, let's close this one and move onto that one. I'm
> actually quite happy with these changes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-core/pull/3152#issuecomment-647500489>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMEOABECADHMC3BUXHW7HDRX5H7VANCNFSM4LRMZI3A>
> .
>

ibacher commented:
Maybe it's better to keep things smaller then. I'll migrate my comment on `birthdate` to here.

ibacher commented on api/src/main/java/org/openmrs/Person.java:
```diff
@@ -59,6 +62,9 @@
 	
+
+	@Field(analyze = Analyze.YES)
+	@DateBridge(encoding = EncodingType.STRING, resolution = Resolution.YEAR)
```
As a general rule, we'd actually want to store the resolution of the birth date as a DAY rather than a YEAR (even for the purposes of getSimilarPeople(), the resolution isn't quite a year, but a fuzzy "3 year period either side of this given year).

ibacher replied:
I can see we do explicitly need a year. If we cannot construct a year query in Lucene, I think we should add a new calculated Lucene-only field of birthyear or something.

fruether replied:
Alright. Then let me work on adding a Lucene only field. I will first do a small research if there is way within Lucene

fruether replied:
@ibacher adding a year field for Lucene sounds reasonable. I will do it that way if you think it is finde

fruether replied:
@ibacher I can set the resolution to DAY and will not end anything on the result.

fruether commented on api/src/main/java/org/openmrs/Person.java:
```diff
@@ -59,6 +62,9 @@
 	
+
+	@Field(analyze = Analyze.YES)
+	@DateBridge(encoding = EncodingType.STRING, resolution = Resolution.DAY)
```
@ibacher I changed it to DAY and it is still working as expected. Is this concern of yours then adressed?

fruether commented:
@ibacher any more concerns regarding this PR?

fruether commented:
@ibacher because I am starting to forget what I have done in march. So it would be great if we could get it through the door soon.
